doc/rrdbotd.8
.\" 
.\" Copyright (c) 2006, Stefan Walter
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without 
.\" modification, are permitted provided that the following conditions 
.\" are met:
.\" 
.\"     * Redistributions of source code must retain the above 
.\"       copyright notice, this list of conditions and the 
.\"       following disclaimer.
.\"     * Redistributions in binary form must reproduce the 
.\"       above copyright notice, this list of conditions and 
.\"       the following disclaimer in the documentation and/or 
.\"       other materials provided with the distribution.
.\"     * The names of contributors to this software may not be 
.\"       used to endorse or promote products derived from this 
.\"       software without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
.\" FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
.\" COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
.\" BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
.\" OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED 
.\" AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, 
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF 
.\" THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH 
.\" DAMAGE.
.\" 
.\"
.\" CONTRIBUTORS
.\"  Stefan Walter <stef@memberwebs.com>
.\"
.Dd August, 2006
.Dt rrdbotd 8
.Os rrdbot 
.Sh NAME
.Nm rddbotd
.Nd an SNMP poller which writes to an round robin database
.Sh SYNOPSIS
.Nm
.Op Fl M
.Op Fl b Ar bindaddr
.Op Fl c Ar confdir
.Op Fl w Ar workdir
.Op Fl m Ar mibdir
.Op Fl d Ar debuglevel
.Op Fl p Ar pidfile
.Op Fl r Ar retries
.Op Fl t Ar timeout
.Nm 
.Fl V
.Sh DESCRIPTION
.Nm
is an SNMP polling daemon which writes the polled values to an 
.Xr rrdtool 1
RRD database. An can poll many different SNMP sources in an efficient manner.  
.Pp
Table queries are supported, where the OID index of a certain value is not 
known beforehand, or may change regularly. 
.Pp
In addition multiple SNMP agents may be specified for a certain value. If 
one SNMP agent cannot be contacted or errors for some reason, another one 
will be tried.
.Pp
The configuration (eg: SNMP sources, polling intervals) are located in files 
in a directory, with one configuration file per RRD. The format of the 
configuration files are described in:
.Xr rrdbot.conf 5
.Sh OPTIONS
The options are as follows. 
.Bl -tag -width Fl
.It Fl b Ar bindaddr
Address to bind to and send SNMP packets from.
.It Fl c Ar confdir
The directory in which configuration files are stored. See below for info
on the various file locations.
.It Fl d Ar debuglevel
Don't detach from the console and run as a daemon. In addition the 
.Ar debuglevel
argument specifies what level of error messages to display. 0 being 
the least, 4 the most.
.It Fl m Ar mibdir
The directory in which to look for MIB files. The default directory is 
usually sufficient.
.It Fl M
Display MIB parsing warnings.
.It Fl p Ar pidfile
Specifies a location for the a process id file to be written to. This file 
contains the process id of 
.Nm 
and can be used to stop the daemon.
.It Fl r Ar retries
The number of times to retry sending an SNMP packet. Defaults to 3 retries.
.It Fl t Ar timeout
The amount of time (in seconds) to wait for an SNMP response. Defaults to 
5 seconds.
.It Fl V
Prints the version of
.Nm
and the locations of the configuration files, RRD files etc.
.It Fl w Ar workdir
The default directory where to look for RRD files. See below for info on 
the various file locations.
.El
.Sh FILE LOCATIONS
To determine the default location for the configuration files and RRD files 
run this command:
.Bd -literal -offset indent
# rrdbotd -V 
.Ed
.Pp
The configuration files for SNMP pollers are laid out in a directory tree, 
with one file per RRD. Subdirectories can be used to organize the 
configuration files. The contents of the configuration files are described 
in 
.Xr rrdbot.conf 5 .
.Pp
By default the RRD files mirror the directory structure and names of the 
configuration files, with an 
.Pa .rrd
extension appended to the filename.
.Pp
For example if your configuration files are in a structure like the following:
.Bd -literal -offset indent
/usr/local/etc/rrdbot/
  gateways/
    gateway-load.conf
    gateway-traffic.conf
  temperature/
    inside-temperature.conf
    outside-temperature.conf
  machine-load.conf  
.Ed
.Pp
Then the default RRD files would be in a similar directory structure:
.Bd -literal -offset indent
/var/db/rrdbot/
  gateways/
    gateway-load.conf.rrd
    gateway-traffic.conf.rrd
  temperature/
    inside-temperature.conf.rrd
    outside-temperature.conf.rrd
  machine-load.conf.rrd
.Ed
.Pp
The default location for an RRD file can be overridden by using the 
.Ar rrd
option in the configuration file.
.Sh SEE ALSO
.Xr rrdbot.conf 5 ,
.Xr rrdbot-get 1 ,
.Xr rrdtool 1
.Sh AUTHOR
.An Stef Walter Aq stef@memberwebs.com


NEWS
0.9.7  [2011-01-11]
    - Fix problems with rollover of SNMP ids.
    - Add real bsnmp-pcap MIB.

0.9.6  [2009-10-29]
    - Fix build problems on FreeBSD.
    - Allow failure for listening on sockets, as long as one
      listen socket works.

0.9.5  [2009-10-28]
    - Move development to git.
    - Add more MIB files from Freebsd and BSNMP
    - Support IPv6 addresses in urls and requests via IPv6.
    - Builds on Solaris
    - Fix memory leaks
    - Better config file parsing.

0.9.4  [2009-01-19]
    - Add support for creating multiple CFs in rrdbot-create. [Stoned Elipot]

0.9.3  [2009-01-16]
    - Add support for printing out MIB name in rrdbot-get

0.9.2  [2008-09-25]
    - Add parameter to specify bind address.

0.9.1  [2008-07-07]
    - Fix bug with table queries in rrdbot-get
    - rrdbot-get now prints out unsigned integers properly.

0.9  [2008-04-16]
    - Support failover between multiple agents.
    - Support table queries by looking up table indexes.
    - Major refactoring of internals.
    - Warn and ignore MIBs that we don't recognize.
    - Support parsing strings into integers and floats


0.8  [2007-07-16]
    - Fix assertion when reallocating SNMP request memory

0.7
    - Use my real name 'Stef Walter'
      See: http://memberwebs.com/nielsen/

0.6
    - Fix bug with asynchronous resolver when daemonizing

0.5
    - Added SNMP version 2c support for rrdbot-get [Stoned Elipot]
    - Fixed asynchronous resolving on FreeBSD
    - Fix bug where rrdbot-create ignored '-c' argument [benj]
    - Remove pid file when exiting [benj]
    - Fix bug where config files are not enumerated on reiserfs [benj]

0.4
    - Reduced memory usage of SNMP requests
    - Added asynchronous DNS resolver
    - Added support for RRD files not in the normal places
    - Added SNMP version 2c support

0.3
    - Initial public release


daemon/rrdbotd.c
/*
 * Copyright (c) 2005, Stefan Walter
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *     * Redistributions of source code must retain the above
 *       copyright notice, this list of conditions and the
 *       following disclaimer.
 *     * Redistributions in binary form must reproduce the
 *       above copyright notice, this list of conditions and
 *       the following disclaimer in the documentation and/or
 *       other materials provided with the distribution.
 *     * The names of contributors to this software may not be
 *       used to endorse or promote products derived from this
 *       software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
 * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 *
 *
 * CONTRIBUTORS
 *  Stef Walter <stef@memberwebs.com>
 *
 */

#include "usuals.h"

#include "async-resolver.h"
#include "log.h"
#include "rrdbotd.h"
#include "server-mainloop.h"
#include "snmp-engine.h"

#include <errno.h>
#include <unistd.h>
#include <stdarg.h>
#include <syslog.h>
#include <signal.h>
#include <err.h>

#include <bsnmp/asn1.h>
#include <bsnmp/snmp.h>
#include <mib/mib-parser.h>


/* The default command line options */
#define DEFAULT_CONFIG      CONF_PREFIX "/rrdbot"
#define DEFAULT_WORK        "/var/db/rrdbot"
#define DEFAULT_RETRIES     3
#define DEFAULT_TIMEOUT     5

/* -----------------------------------------------------------------------------
 * GLOBALS
 */

/* The one main state object */
rb_state g_state;

/* Some logging flags */
static int daemonized = 0;
static int debug_level = LOG_ERR;

#include "mib/parse.h"

#ifdef TEST

static void
test(int argc, char* argv[])
{
    struct snmp_value val;

    debug_level = 4;

    if(argc < 2)
        errx(2, "specify arguments");

    while(argc > 1)
    {
        if(rb_snmp_parse_mib(argv[1], &val) == -1)
            warnx("couldn't parse mib value: %s", argv[1]);
        else
            fprintf(stderr, "the oid is: %s\n", asn_oid2str(&(val.var)));

        argc--;
        argv++;
    }

    rb_mib_uninit();
    exit(1);
}

#endif /* TEST */

/* -----------------------------------------------------------------------------
 * LOGGING
 */

void
log_vmessage(int level, int erno, const char* msg, va_list ap)
{
    #define MAX_MSGLEN  1024
    char buf[MAX_MSGLEN];

    if(daemonized) {
        if (level >= LOG_DEBUG)
            return;
    } else {
        if(debug_level < level)
            return;
    }

    ASSERT (msg);

    /* Cleanup the message a little */
    strlcpy(buf, msg, MAX_MSGLEN);
    stretrim(buf);

    if(erno)
    {
        strlcat(buf, ": ", MAX_MSGLEN);
        strncat(buf, strerror(erno), MAX_MSGLEN);
    }

    /* As a precaution */
    buf[MAX_MSGLEN - 1] = 0;

    /* Either to syslog or stderr */
    if(daemonized)
        vsyslog(level, buf, ap);
    else
        vwarnx(buf, ap);
}

/* -----------------------------------------------------------------------------
 * STARTUP
 */

static void
usage()
{
    fprintf(stderr, "usage: rrdbotd [-M] [-c confdir] [-w workdir] [-m mibdir] \n");
    fprintf(stderr, "               [-d level] [-p pidfile] [-r retries] [-t timeout]\n");
    fprintf(stderr, "       rrdbotd -V\n");
    exit(2);
}

static void
version()
{
    printf("rrdbotd (version %s)\n", VERSION);
    printf("   default config directory: %s\n", DEFAULT_CONFIG);
    printf("   default work directory:   %s\n", DEFAULT_WORK);
    printf("   default mib directory:    %s\n", DEFAULT_MIB);
    exit(0);
}

static void
on_quit(int signal)
{
    fprintf(stderr, "rrdbotd: got signal to quit\n");
    server_stop();
}

static void
writepid(const char* pidfile)
{
    FILE* f = fopen(pidfile, "w");
    if(f == NULL)
    {
        log_warn("couldn't open pid file: %s", pidfile);
        return;
    }

    fprintf(f, "%d\n", (int)getpid());
    if(ferror(f))
        log_warn("couldn't write to pid file: %s", pidfile);

    fclose(f);
}

static void
removepid(const char* pidfile)
{
    if(unlink(pidfile) < 0)
    {
        if(errno != ENOENT)
            log_warn("couldn't remove pid file: %s", pidfile);
    }
}

int
main(int argc, char* argv[])
{
	const char** local = NULL;
	int n_local = 0;
    const char* pidfile = NULL;
    int daemonize = 1;
    char ch;
    char* t;

#ifdef TEST
    test(argc, argv);
    return 1;
#endif

    /* Initialize the state stuff */
    memset(&g_state, 0, sizeof(g_state));

    g_state.rrddir = DEFAULT_WORK;
    g_state.confdir = DEFAULT_CONFIG;
    g_state.retries = DEFAULT_RETRIES;
    g_state.timeout = DEFAULT_TIMEOUT;

    /* Parse the arguments nicely */
    while((ch = getopt(argc, argv, "b:c:d:m:Mp:r:t:w:V")) != -1)
    {
        switch(ch)
        {

        /* Bind address */
        case 'b':
            local = xrealloc (local, sizeof (char*) * (n_local + 2));
            local[n_local] = optarg;
            local[++n_local] = NULL;
            break;

        /* Config directory */
        case 'c':
            g_state.confdir = optarg;
            break;

        /* Don't daemonize */
        case 'd':
            daemonize = 0;
            debug_level = strtol(optarg, &t, 10);
            if(*t || debug_level > 4)
                errx(1, "invalid debug log level: %s", optarg);
            debug_level += LOG_ERR;
            break;

        /* mib directory */
        case 'm':
            mib_directory = optarg;
            break;

        /* MIB load warnings */
        case 'M':
            mib_warnings = 1;
            break;

        /* Write out a pid file */
        case 'p':
            pidfile = optarg;
            break;

        /* The number of SNMP retries */
        case 'r':
            g_state.retries = strtol(optarg, &t, 10);
            if(*t || g_state.retries < 0)
                errx(1, "invalid number of retries: %s", optarg);
            break;

        /* The default timeout */
        case 't':
            g_state.timeout = strtol(optarg, &t, 10);
            if(*t || g_state.timeout <= 0)
                errx(1, "invalid timeout (must be above zero): %s", optarg);
            break;

        /* The work directory */
        case 'w':
            g_state.rrddir = optarg;
            break;

        /* Print version number */
        case 'V':
            version();
            break;

        /* Usage information */
        case '?':
        default:
            usage();
            break;
        }
    }

    argc -= optind;
    argv += optind;

    if(argc != 0)
        usage();

    /* No bind addresses specified, use defaults... */
    if (local == NULL) {
        local = xrealloc (local, sizeof (char*) * 3);
        local[0] = "0.0.0.0";
        local[1] = NULL;
#ifdef HAVE_INET6
        local[1] = "::";
        local[2] = NULL;
#endif
    }

    /* The mainloop server */
    server_init();

    /* Parse config and setup SNMP system */
    rb_config_parse();

    /* As an optimization we unload the MIB processing data here */
    mib_uninit();

    /* Rev up the main engine */
    snmp_engine_init (local, g_state.retries);
    rb_poll_engine_init();

    free (local);
    n_local = 0;
    local = NULL;

    if(daemonize)
    {
        /* Fork a daemon nicely */
        if(daemon(0, 0) == -1)
            err(1, "couldn't fork as a daemon");

        log_debug("running as a daemon");
        daemonized = 1;
    }

    /* Setup the Async DNS resolver */
    if(async_resolver_init() < 0)
    {
        log_error("couldn't initialize resolver");
        /* Allow things to proceed without resolver */
    }

    /* Handle signals */
    signal(SIGPIPE, SIG_IGN);
    signal(SIGHUP, SIG_IGN);
    signal(SIGINT,  on_quit);
    signal(SIGTERM, on_quit);
    siginterrupt(SIGINT, 1);
    siginterrupt(SIGTERM, 1);

    /* Open the system log */
    openlog("rrdbotd", 0, LOG_DAEMON);

    if(pidfile != NULL)
        writepid(pidfile);

    log_info("rrdbotd version " VERSION " started up");

    /* Now let it go */
    if(server_run() == -1)
        err(1, "critical failure running SNMP engine");

    log_info("rrdbotd stopping");

    /* Cleanups */
    rb_poll_engine_uninit();
    snmp_engine_stop();
    rb_config_free();
    async_resolver_uninit();
    server_uninit();

    if(pidfile != NULL)
        removepid(pidfile);

    return 0;
}
